File: projects/grind-75/grind-75-python/graph/clone_graph.py
from typing import Optional
from graph.common import Node


class Solution:
    @classmethod
    def cloneGraph(self, node: Optional[Node]) -> Optional[Node]:
        hmap = {}

        def dfs(node):
            if node in hmap:
                return hmap[node]
            copy = Node(node.val)
            hmap[node] = copy
            for neighbor in node.neighbors:
                copy.neighbors.append(dfs(neighbor))
            return copy

        return dfs(node) if node else None


"""
Time Complexity - O(V+E)
Space Complexity - O(V)
"""

File: projects/grind-75/typescript/graph/clone-graph.ts
import { _Node } from './common';

export function cloneGraph(node: _Node | null): _Node | null {
  const store = new Map<number, _Node>();
  function dfs(node: _Node): _Node {
    if (store.has(node.val)) {
      return store.get(node.val) as _Node;
    }
    const copy = new _Node(node.val);
    store.set(node.val, copy);
    for (let neighbor of node.neighbors) {
      copy.neighbors.push(dfs(neighbor));
    }
    return copy;
  }
  return node ? dfs(node) : null;
}
